collector/slabinfo_linux.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build linux && !noslabinfo
+// +build linux,!noslabinfo
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+type slabinfoCollector struct {
+	fs     procfs.FS
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("slabinfo", defaultDisabled, NewSlabinfoCollector)
+}
+
+func NewSlabinfoCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+
+	return &slabinfoCollector{logger: logger, fs: fs}, nil
+}
+
+func (c *slabinfoCollector) Update(ch chan<- prometheus.Metric) error {
+	slabinfo, err := c.fs.SlabInfo()
+	if err != nil {
+		return fmt.Errorf("couldn't get slabinfo: %w", err)
+	}
+
+	for _, slab := range slabinfo.Slabs {
+		// To avoid confusion, the naming conventions for metrics
+		// are taken directly from code of Linux Kernel.
+		ch <- slabMetric("active_objs", slab.Name, float64(slab.ObjActive))
+		ch <- slabMetric("num_objs", slab.Name, float64(slab.ObjNum))
+		ch <- slabMetric("objsize", slab.Name, float64(slab.ObjSize))
+		ch <- slabMetric("objperslab", slab.Name, float64(slab.ObjPerSlab))
+		ch <- slabMetric("pagesperslab", slab.Name, float64(slab.PagesPerSlab))
+	}
+	return nil
+}
+
+func slabMetric(key, label string, val float64) prometheus.Metric {
```

Comment:
Since we have relatively few metric names, I would prefer to have them created separately. This way we can also have the help strings from the [slabinfo man paage](https://man7.org/linux/man-pages/man5/slabinfo.5.html).

Reply:
Thanks for the review.
I have split the function and improved metric names and help strings.